src/datacollector/DataCollectionAttachmentManager.cs

```diff
@@ -0,0 +1,458 @@
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+namespace Microsoft.VisualStudio.TestPlatform.DataCollector
+{
+    using System;
+    using System.Collections.Generic;
+    using System.ComponentModel;
+    using System.Diagnostics;
+    using System.Globalization;
+    using System.IO;
+    using System.Linq;
+
+    using Microsoft.VisualStudio.TestPlatform.DataCollector.Interfaces;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.DataCollection;
+    using Microsoft.VisualStudio.TestPlatform.ObjectModel.Logging;
+    using Microsoft.VisualStudio.TestPlatform.Utilities;
+
+    /// <summary>
+    /// Manages file transfer from data collector to test runner service.
+    /// </summary>
+    internal class DataCollectionAttachmentManager : IDataCollectionAttachmentManager
+    {
+        #region Fields
+
+        /// <summary>
+        /// Max number of file transfer jobs.
+        /// </summary>
+        private const int MaxQueueLength = 500;
+
+        /// <summary>
+        /// Max queue size
+        /// </summary>
+        private const int MaxQueueSize = 25000000;
+
+        /// <summary>
+        /// Default results directory to be used when user didn't specify.
+        /// </summary>
+        private const string DefaultOutputDirectoryName = "TestResults";
+
+        /// <summary>
+        /// Specifies whether the object is disposed or not. 
+        /// </summary>
+        private bool disposed;
+
+        /// <summary>
+        /// Job queue for moving file from source to destination.
+        /// </summary>
+        private JobQueue<AttachmentRequest> fileCopierJobQueue;
+
+        /// <summary>
+        /// Logger for data collection messages
+        /// </summary>
+        private IMessageSink messageSink;
+
+        #endregion
+
+        #region Constructor
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="DataCollectionAttachmentManager"/> class.
+        /// </summary>
+        public DataCollectionAttachmentManager()
+        {
+            this.AttachmentRequests = new List<AttachmentRequest>();
+
+            this.fileCopierJobQueue = new JobQueue<AttachmentRequest>(
+                this.OnAttachmentRequest,
+                "DataCollectionAttachmentManagerQueue",
+                MaxQueueLength,
+                MaxQueueSize,
+                true,
+                message => EqtTrace.Error(message));
+        }
+
+        #endregion
+
+        #region Properties
+
+        /// <summary>
+        /// Gets the session output directory.
+        /// </summary>
+        internal string SessionOutputDirectory { get; private set; }
+
+        /// <summary>
+        /// Gets the list of attachment requests.
+        /// </summary>
+        internal List<AttachmentRequest> AttachmentRequests { get; private set; }
```

Comment:
>AttachmentRequests [](start = 41, length = 18)

It should be possible to replace this with async Tasks for Copy/Move. We could wait on Task.WhenAll(attachmentRequests)?